### DISCUSSION TITLE
Storing secrets in the process env variables is risky

### DISCUSSION THREAD

**Discussion by eladchen** [motdotla/dotenv#752] [Ideas]
I really like the solution offered by dotenv vault, just not as advertised.

Storing secrets in the process env variables is risky, it allows 3rd party modules
which run on the same process access to read the secrets..

As an architect in a tech company I'm considering using this tool like so:
- Create a js script that **DOES NOT** use any 3rd party modules
- Load the env variables using dotenv
- Flush the env variables needed by the app to a file and exit

This will result in a file with secrets existing on the disk.

The other part would be to start the actual process, init a module which will read the file and
delete it from the disk, thus it cannot be accessed by a malicious 3rd party module, whilst ensuring the secrets
are only available via dedicated module api.

WDYT?

**Comment by motdotla** [comment]
> This will result in a file with secrets existing on the disk.

This has one downside. For server environments that have ephemeral file storage (like Heroku and many others) you can't guarantee the file will always be there. For other deployment environments like serverless, you can't always write to a file. They don't allow it.

Instead, maintaining the encrypted file in your codebase guarantees it will be available to all infra types.

> Storing secrets in the process env variables is risky; it allows 3rd party modules which run on the same process access to read the secrets

This is a good thought. They only exist during that process (and are cleared when the process shuts down), but yes, if a 3rd party accesses that process somehow, then it can slurp them up. This is rare and means you have a worse intrusion (like a compromised node module or server), but I agree with the point.

What if we added the ability for you to pass your own target that should be written to - rather than `process.env`?

https://github.com/motdotla/dotenv/blob/master/lib/main.js#L188

We recently built the mechanism to do to this (https://github.com/motdotla/dotenv/tree/master#populate). We would just need to bubble up the option to `config`. Something like this:

```
const target = {}
require('dotenv').config(target: target)

console.log(target) // values from .env or .env.vault live here now.
console.log(process.env) // this was not changed or written to
```

What do you think?

**Reply by eladchen** [reply]
Sounds good!

I love that idea, that would be perfect. I guess it time for a POC then.

**Reply by motdotla** [reply]
Coming together here: https://github.com/motdotla/dotenv/pull/753/files

**Reply by motdotla** [reply]
Available as `16.2.0`

**Reply by motdotla** [reply]
Here's how to use it: https://github.com/motdotla/dotenv#processenv

```
const myObject = {}
require('dotenv').config({ processEnv: myObject })

console.log(myObject) // values from .env or .env.vault live here now.
console.log(process.env) // this was not changed or written to
```

**Reply by eladchen** [reply]
What about the DOTENV_KEY? Having to set it when running the process is just as risky..

Any ideas?

**Reply by motdotla** [reply]
We could do similar. Pass it as an optional option? Otherwise defaults to process.env.

**Reply by eladchen** [reply]
Sounds good. Always a good idea to have a library API to be as direct as possible, also makes testing a lot easier.

**Reply by eladchen** [reply]
BTW

You said:

"For server environments that have ephemeral file storage (like Heroku and many others) you can't guarantee the file will always be there."

Ephemeral storage or not, the process "boot sequence" was what I was suggesting, where the
file with the secrets I suggested is generated somehow during boot.

**Reply by motdotla** [reply]
> where the file with the secrets I suggested is generated somehow during boot.

After this feature is released, I'd like to see how you use it that way. My initial thought is it seems redundant - since the .env.vault file will already be there committed to your code - but I'd like to be educated on it further.

**Reply by eladchen** [reply]
boot.js
```js
const fs = require("fs/promises");
const { spawn } = require("child_process");

const loadEnv = (target = {}) => {
  // Make sure this is commented out
  // require("dotenv").config({ processEnv: target });

  // For example:
  target.db_secret = "test";

  return target;
};

const writeEnvToSecretsFile = async (env) => {
  // The content for the ES module with the secrets
  const content = [
    '// Generated file. Do not edit.',
    `module.exports = ${JSON.stringify(env, null, 2)};`,
  ];

  await fs.writeFile("./secrets.js", content.join("\n"));
};

writeEnvToSecretsFile(loadEnv()).then(() => {
  console.log("Great success");

  // Two options here, spawn a new proces / let this process die
  // and execute another node js script...

  // An example which spawns a new process to replace the current process
  const newProcess = spawn("node app.js", [], {
    stdio: "inherit",
    shell: true,
  });

  newProcess.on("exit", (code, _signal) => {
    process.exit(code);
  });
});
```

app.js
```js
const secrets = require("./secrets.js");

console.log("secrets file", secrets);
console.log("env variables", process.env);
```

I wrote this quick and dirty, but it seems to be working and what I aim for..

Perhaps dotenv could have an API which accepts a spawn callback it can invoke?

```js
require("dotenv").config({
  async spawn(envVariables) {
    // This is where power users can do what I did above:
    // Flush the variable to disk
    // Start a new process?
  }
});
```

**Reply by eladchen** [reply]
P.S The app.js or even dotenv itself could delete the file one the new process spawnd, this will ensure no malicious 3rd party modules scans/dumps the file system

**Reply by motdotla** [reply]
Ok, interesting. With `16.3.0` (just released) supporting `options.DOTENV_KEY`, you could optionally do this:

```
const result = require('dotenv').config({ processEnv: {}, DOTENV_KEY: 'dotenv://:key_cd9fe909e1adf109873d5117c3f4beda33a9a8006f4eaeb73a2a0d21bab0cbd4@dotenv.local/vault/.env.vault?environment=development' })

console.log(result.parsed) // { HELLO: 'Universe' }
console.log(result.parsed.HELLO) // 'Universe'
console.log(process.env.HELLO) // undefined (as intentioned)
```

You could still then write the `result.parsed` object to a `secrets.js` file.

What do you think?

[1] https://github.com/dotenv-org/examples/blob/master/dotenv-vault-custom-target/index.js

**Reply by eladchen** [reply]
I can write the file using dotenv with/without using special config.. The idea is is separating the boot process into two phases
the phase where secrets are retrieved, and the phase where the app boots and reads a file with the secrets.

I really think processEnv is a good option, but it only solves half the problem given DOTENV_KEY is still needed.

If you ask me, I think two options should exists, "processEnv", and "spawn" - a callback where the process boot can take place.
When one option exists, the other is ignored.

I would have loved it if the docs outlined this advanced usage and the added security benefits it provides.

**Reply by eladchen** [reply]
Even if the spawn option seems a bit much, I would still integrate it the way I showed, just because I can & believe its more secured.

Just wanted to share my ideas. 🙃

**Reply by motdotla** [reply]
@eladchen in this case by spawning that process, those environment variables are technically still available to that process and any third-party module that would have access to that process - they are just in a different place, adding a bit more friction for an attacking third-party. So I think that can be useful in your case.

BUT

I want to clarify that the dotenv module does NOT permanently set environment variables on your machine in case you are under that impression. It sets them only temporarily for the process it is run in. For example, if a process runs and then quits, the values set by dotenv do not persist. A second, or third, or fourth process cannot gain access to them unless they also somehow have access to the same process that dotenv was called in.
